core/src/main/java/org/infinispan/util/concurrent/jdk8backported/StrippedConcurrentLinkedDeque.java

```diff
@@ -0,0 +1,723 @@
+               itemOffset = UNSAFE.objectFieldOffset
+                   (k.getDeclaredField("item"));
+               nextOffset = UNSAFE.objectFieldOffset
+                   (k.getDeclaredField("next"));
+           } catch (Exception e) {
+               throw new Error(e);
+           }
+       }
+   }
+
+   abstract class AbstractItr implements Iterator<DequeNode<E>> {
+      /**
+       * Next node to return item for.
+       */
+      private DequeNode<E> nextNode;
+
+      /**
+       * Node returned by most recent call to next. Needed by remove.
+       * Reset to null if this element is deleted by a call to remove.
+       */
+      private DequeNode<E> lastRet;
+
+      abstract DequeNode<E> startNode();
+      abstract DequeNode<E> nextNode(DequeNode<E> p);
+
+      AbstractItr() {
+          advance();
+      }
+
+      /**
+       * Sets nextNode and nextItem to next valid node, or to null
+       * if no such.
+       */
+      private void advance() {
+          lastRet = nextNode;
+
+          DequeNode<E> p = (nextNode == null) ? startNode() : nextNode(nextNode);
+          for (;; p = nextNode(p)) {
+              if (p == null) {
+                  // p might be active end or TERMINATOR node; both are OK
+                  nextNode = null;
+                  break;
+              }
+              E item = p.item;
+              if (item != null) {
+                  nextNode = p;
+                  break;
+              }
+          }
+      }
+
+      public boolean hasNext() {
+          return nextNode != null;
+      }
+
+      public DequeNode<E> next() {
+          DequeNode<E> node = nextNode;
+          if (node == null) throw new NoSuchElementException();
+          advance();
+          return node;
+      }
+
+      public void remove() {
+         DequeNode<E> l = lastRet;
+          if (l == null) throw new IllegalStateException();
+          l.item = null;
+          unlink(l);
+          lastRet = null;
+      }
+  }
+
+  /** Forward iterator */
+  class Itr extends AbstractItr {
+      DequeNode<E> startNode() { return first(); }
+      DequeNode<E> nextNode(DequeNode<E> p) { return succ(p); }
+  }
+
+  /** Descending iterator */
+  class DescendingItr extends AbstractItr {
+     DequeNode<E> startNode() { return last(); }
+     DequeNode<E> nextNode(DequeNode<E> p) { return pred(p); }
+  }
+
+   public E pollFirst() {
+      for (DequeNode<E> p = first(); p != null; p = succ(p)) {
+          E item = p.item;
+          if (item != null && p.casItem(item, null)) {
+              unlink(p);
+              return item;
+          }
+      }
+      return null;
+   }
+
+   /**
+    * Returns the first node as well as it's value
+    * @return an array of size 2 (1 = DequeNode<E> that was removed (note item will be 
+    *         null, and 2 = value of the node from 1).  Or this will return null if there
+    *         are no nodes in this deque
+    */
+   public Object[] pollFirstNode() {
```

Comment:
maybe... since the method is invoked in a loop, what about change it to receive the ```Object[]```? It would avoid creating an array for each invocation. 

```boolean pollFirstNode(Object[] result) //returns true if a node is found and the array was populated, false otherwise.```